File: ontology/scripts/generate.ts
#!/usr/bin/env tsx
/**
 * PCOMJR TypeScript Code Generator
 * 
 * Reads schema/pcomjr.schema.json and emits:
 *   - ontology/src/generated/types.ts — All interfaces, type aliases, enums
 *   - ontology/src/generated/guards.ts — Type guard functions
 *
 * Run: cd pcomjr-spec && npx tsx ontology/scripts/generate.ts
 */

import { readFileSync, writeFileSync, mkdirSync } from 'fs';
import { resolve, dirname } from 'path';

// ─── Schema Loading ─────────────────────────────────────────────────────────

const SCHEMA_PATH = resolve(import.meta.dirname, '../../schema/pcomjr.schema.json');
const OUT_DIR = resolve(import.meta.dirname, '../src/generated');

interface SchemaDef {
  type?: string;
  enum?: string[];
  const?: string;
  description?: string;
  required?: string[];
  properties?: Record<string, any>;
  items?: any;
  allOf?: any[];
  oneOf?: any[];
  $ref?: string;
  additionalProperties?: any;
  format?: string;
  minimum?: number;
  maximum?: number;
  minLength?: number;
  minItems?: number;
  pattern?: string;
  default?: any;
}

const schema = JSON.parse(readFileSync(SCHEMA_PATH, 'utf8'));
const defs: Record<string, SchemaDef> = schema.$defs;

// ─── Classification ─────────────────────────────────────────────────────────

// Types that are string enums
const ENUM_TYPES = new Set<string>();
// Types that are union refs (oneOf with $ref entries)
const UNION_TYPES = new Set<string>();
// Types that are object interfaces
const INTERFACE_TYPES = new Set<string>();
// Types that extend Artifact (have allOf with $ref to Artifact)
const ARTIFACT_SUBTYPES = new Set<string>();

for (const [name, def] of Object.entries(defs)) {
  if (def.type === 'string' && def.enum) {
    ENUM_TYPES.add(name);
  } else if (def.oneOf && def.oneOf.every((o: any) => o.$ref)) {
    UNION_TYPES.add(name);
  } else if (def.type === 'object' || def.allOf) {
    INTERFACE_TYPES.add(name);
    if (def.allOf?.some((a: any) => a.$ref === '#/$defs/Artifact')) {
      ARTIFACT_SUBTYPES.add(name);
    }
  }
}

// ─── Type Resolution ────────────────────────────────────────────────────────

function resolveRef(ref: string): string {
  return ref.replace('#/$defs/', '');
}

function schemaToTS(prop: any, optional: boolean = false): string {
  if (prop.$ref) {
    return resolveRef(prop.$ref);
  }

  if (prop.const) {
    return JSON.stringify(prop.const);
  }

  if (prop.oneOf) {
    return prop.oneOf.map((o: any) => schemaToTS(o)).join(' | ');
  }

  if (prop.type === 'string') {
    if (prop.enum) {
      return prop.enum.map((v: string) => JSON.stringify(v)).join(' | ');
    }
    return 'string';
  }

  if (prop.type === 'number' || prop.type === 'integer') {
    return 'number';
  }

  if (prop.type === 'boolean') {
    return 'boolean';
  }

  if (prop.type === 'array') {
    if (prop.items) {
      const itemType = schemaToTS(prop.items);
      // If items is a union type (contains |), wrap in parens before []
      const needsParens = itemType.includes(' | ');
      return needsParens ? `(${itemType})[]` : `${itemType}[]`;
    }
    return 'unknown[]';
  }

  if (prop.type === 'object') {
    if (prop.properties && Object.keys(prop.properties).length > 0) {
      // Inline object with known properties
      const fields = Object.entries(prop.properties).map(([k, v]: [string, any]) => {
        const req = prop.required?.includes(k) ?? false;
        return `${k}${req ? '' : '?'}: ${schemaToTS(v as any)}`;
      });
      return `{ ${fields.join('; ')} }`;
    }
    if (prop.additionalProperties === true || (prop.additionalProperties && typeof prop.additionalProperties === 'object')) {
      const valueType = typeof prop.additionalProperties === 'object' && prop.additionalProperties.type
        ? schemaToTS(prop.additionalProperties)
        : 'unknown';
      return `Record<string, ${valueType}>`;
    }
    return 'any';
  }

  return 'unknown';
}

// ─── Code Generation ────────────────────────────────────────────────────────

function generateTypes(): string {
  const lines: string[] = [];

  lines.push('/**');
  lines.push(' * PCOMJR Data Ontology — Generated TypeScript Types');
  lines.push(' * ');
  lines.push(' * AUTO-GENERATED from schema/pcomjr.schema.json');
  lines.push(' * DO NOT EDIT MANUALLY — run `npm run generate` to regenerate.');
  lines.push(` * Generated: ${new Date().toISOString()}`);
  lines.push(' */');
  lines.push('');

  // ── Enum types as string literal unions ──
  lines.push('// ═══════════════════════════════════════════════════════════════');
  lines.push('// String Literal Types (Enums)');
  lines.push('// ═══════════════════════════════════════════════════════════════');
  lines.push('');

  for (const name of sorted(ENUM_TYPES)) {
    const def = defs[name];
    if (def.description) lines.push(`/** ${def.description} */`);
    const values = def.enum!.map(v => JSON.stringify(v)).join(' | ');
    lines.push(`export type ${name} = ${values};`);
    lines.push('');

    // Also export the values as a const array for runtime use
    const constName = toScreamingSnake(name);
    lines.push(`export const ${constName} = [${def.enum!.map(v => JSON.stringify(v)).join(', ')}] as const;`);
    lines.push('');
  }

  // ── Union types ──
  lines.push('// ═══════════════════════════════════════════════════════════════');
  lines.push('// Union Types');
  lines.push('// ═══════════════════════════════════════════════════════════════');
  lines.push('');

  for (const name of sorted(UNION_TYPES)) {
    const def = defs[name];
    if (def.description) lines.push(`/** ${def.description} */`);
    const members = def.oneOf!.map((o: any) => resolveRef(o.$ref)).join(' | ');
    lines.push(`export type ${name} = ${members};`);
    lines.push('');
  }

  // ── Simple object interfaces (no allOf) ──
  lines.push('// ═══════════════════════════════════════════════════════════════');
  lines.push('// Interfaces');
  lines.push('// ═══════════════════════════════════════════════════════════════');
  lines.push('');

  // Emit non-artifact interfaces first
  const simpleInterfaces = sorted(INTERFACE_TYPES).filter(n => !ARTIFACT_SUBTYPES.has(n) && n !== 'Decision');
  for (const name of simpleInterfaces) {
    const def = defs[name];
    emitInterface(lines, name, def);
  }

  // ── Artifact subtypes ──
  lines.push('// ═══════════════════════════════════════════════════════════════');
  lines.push('// CIAS Artifact Subtypes');
  lines.push('// ═══════════════════════════════════════════════════════════════');
  lines.push('');

  for (const name of sorted(ARTIFACT_SUBTYPES)) {
    if (name === 'Decision') continue; // Decision gets its own section below
    const def = defs[name];
    emitArtifactSubtype(lines, name, def);
  }

  // ── Decision (special case — allOf with extra required) ──
  lines.push('// ═══════════════════════════════════════════════════════════════');
  lines.push('// Decision (extends Artifact)');
  lines.push('// ═══════════════════════════════════════════════════════════════');
  lines.push('');

  const decisionDef = defs['Decision'];
  emitDecision(lines, decisionDef);

  return lines.join('\n');
}

function emitInterface(lines: string[], name: string, def: SchemaDef): void {
  if (def.description) lines.push(`/** ${def.description} */`);
  lines.push(`export interface ${name} {`);
  
  const required = new Set(def.required || []);
  const props = def.properties || {};
  
  for (const [propName, propDef] of Object.entries(props)) {
    const tsType = schemaToTS(propDef);
    const opt = required.has(propName) ? '' : '?';
    if ((propDef as any).description) {
      lines.push(`  /** ${(propDef as any).description} */`);
    }
    lines.push(`  ${propName}${opt}: ${tsType};`);
  }
  
  lines.push('}');
  lines.push('');
}

function emitArtifactSubtype(lines: string[], name: string, def: SchemaDef): void {
  if (def.description) lines.push(`/** ${def.description} */`);
  lines.push(`export interface ${name} extends Artifact {`);
  
  // Collect all properties from allOf entries (excluding the Artifact ref)
  const allProps: Record<string, any> = {};
  const allRequired = new Set<string>();
  
  // The top-level properties and required
  if (def.properties) Object.assign(allProps, def.properties);
  if (def.required) def.required.forEach(r => allRequired.add(r));
  
  // Properties from allOf (skip the $ref to Artifact)
  if (def.allOf) {
    for (const entry of def.allOf) {
      if (entry.$ref) continue; // Skip Artifact base ref
      if (entry.properties) Object.assign(allProps, entry.properties);
      if (entry.required) entry.required.forEach((r: string) => allRequired.add(r));
    }
  }
  
  // Remove properties that are on the base Artifact type (they're inherited)
  const artifactProps = new Set(Object.keys(defs['Artifact'].properties || {}));
  // But keep overridden properties (like type: const)
  
  for (const [propName, propDef] of Object.entries(allProps)) {
    // Skip inherited props unless they have a const override or narrow the type ($ref)
    if (artifactProps.has(propName) && !propDef.const && !propDef.$ref) continue;
    
    const tsType = schemaToTS(propDef);
    // For the subtype, Artifact-level required fields are already on base
    // But content is always required if it has a $ref (type narrowing)
    const isRequired = allRequired.has(propName) && !artifactProps.has(propName) || (propName === 'content' && propDef.$ref);
    const opt = isRequired || propDef.const ? '' : '?';
    if (propDef.description) {
      lines.push(`  /** ${propDef.description} */`);
    }
    lines.push(`  ${propName}${opt}: ${tsType};`);
  }
  
  lines.push('}');
  lines.push('');
}

function emitDecision(lines: string[], def: SchemaDef): void {
  if (def.description) lines.push(`/** ${def.description} */`);
  lines.push('export interface Decision extends Artifact {');
  
  // Merge properties from allOf
  const allProps: Record<string, any> = {};
  const allRequired = new Set<string>();
  
  if (def.allOf) {
    for (const entry of def.allOf) {
      if (entry.$ref) continue;
      if (entry.properties) Object.assign(allProps, entry.properties);
      if (entry.required) entry.required.forEach((r: string) => allRequired.add(r));
    }
  }
  
  const artifactProps = new Set(Object.keys(defs['Artifact'].properties || {}));
  
  for (const [propName, propDef] of Object.entries(allProps)) {
    if (artifactProps.has(propName) && !propDef.const && !propDef.$ref) continue;
    
    const tsType = schemaToTS(propDef);
    const isRequired = allRequired.has(propName) || (propName === 'content' && propDef.$ref);
    const opt = isRequired || propDef.const ? '' : '?';
    if (propDef.description) {
      lines.push(`  /** ${propDef.description} */`);
    }
    lines.push(`  ${propName}${opt}: ${tsType};`);
  }
  
  lines.push('}');
  lines.push('');
}

function generateGuards(): string {
  const lines: string[] = [];

  lines.push('/**');
  lines.push(' * PCOMJR Data Ontology — Generated Type Guards');
  lines.push(' * ');
  lines.push(' * AUTO-GENERATED from schema/pcomjr.schema.json');
  lines.push(' * DO NOT EDIT MANUALLY — run `npm run generate` to regenerate.');
  lines.push(` * Generated: ${new Date().toISOString()}`);
  lines.push(' */');
  lines.push('');
  // Dynamic imports based on schema artifact subtypes
  const subtypeNames = sorted(ARTIFACT_SUBTYPES);
  const allImports = ['Artifact', ...subtypeNames].join(', ');
  lines.push(`import type { ${allImports} } from './types.js';`);
  lines.push("import { CIAS_TERMINAL_ID, OIAS_SYSTEM_ID } from './types.js';");
  lines.push('');

  // isDecision
  lines.push('/** Type guard: is this artifact a Decision? */');
  lines.push('export function isDecision(a: Artifact): a is Decision {');
  lines.push("  return a.type === 'decision';");
  lines.push('}');
  lines.push('');

  // Per-CIAS-type guards — read const values from schema
  const ciasGuards: Array<{ name: string; typeConst: string; iface: string }> = [];
  for (const [defName, defn] of Object.entries(defs)) {
    if (!ARTIFACT_SUBTYPES.has(defName) || defName === 'Decision') continue;
    // Extract the const type value from allOf
    for (const entry of (defn as any).allOf || []) {
      if (entry.$ref) continue;
      const typeConst = entry.properties?.type?.const;
      if (typeConst) {
        const guardName = 'is' + defName;
        ciasGuards.push({ name: guardName, typeConst, iface: defName });
      }
    }
  }
  ciasGuards.sort((a, b) => a.name.localeCompare(b.name));

  for (const { name, typeConst, iface } of ciasGuards) {
    lines.push(`/** Type guard: is this artifact a ${iface}? */`);
    lines.push(`export function ${name}(a: Artifact): a is ${iface} {`);
    lines.push(`  return a.type === '${typeConst}';`);
    lines.push('}');
    lines.push('');
  }

  // isVerifiedArtifact
  lines.push('/** Type guard: has this artifact been verified? */');
  lines.push('export function isVerifiedArtifact(a: Artifact): boolean {');
  lines.push("  return a.verificationStatus === 'verified';");
  lines.push('}');
  lines.push('');

  // isCIASArtifact / isOIASArtifact
  lines.push('/** Type guard: is this artifact from a CIAS terminal? */');
  lines.push('export function isCIASArtifact(a: Artifact): boolean {');
  lines.push('  return (CIAS_TERMINAL_ID as readonly string[]).includes(a.sourceTerminal);');
  lines.push('}');
  lines.push('');

  lines.push('/** Type guard: is this artifact from an OIAS system? */');
  lines.push('export function isOIASArtifact(a: Artifact): boolean {');
  lines.push('  return (OIAS_SYSTEM_ID as readonly string[]).includes(a.sourceTerminal);');
  lines.push('}');
  lines.push('');

  return lines.join('\n');
}

// ─── Helpers ────────────────────────────────────────────────────────────────

function sorted(set: Set<string>): string[] {
  return [...set].sort();
}

function toScreamingSnake(name: string): string {
  // CIASTerminalId -> CIAS_TERMINAL_ID
  // ArtifactType -> ARTIFACT_TYPE
  return name
    .replace(/([a-z])([A-Z])/g, '$1_$2')
    .replace(/([A-Z]+)([A-Z][a-z])/g, '$1_$2')
    .toUpperCase();
}

// ─── Main ───────────────────────────────────────────────────────────────────

mkdirSync(OUT_DIR, { recursive: true });

const typesCode = generateTypes();
const guardsCode = generateGuards();

writeFileSync(resolve(OUT_DIR, 'types.ts'), typesCode);
writeFileSync(resolve(OUT_DIR, 'guards.ts'), guardsCode);

console.log(`✓ Generated ${OUT_DIR}/types.ts (${ENUM_TYPES.size} enums, ${UNION_TYPES.size} unions, ${INTERFACE_TYPES.size} interfaces)`);
console.log(`✓ Generated ${OUT_DIR}/guards.ts`);

File: python/scripts/generate.py
#!/usr/bin/env python3
"""
PCOMJR Python Code Generator

Reads schema/pcomjr.schema.json and emits:
  - python/talonsight_ontology/generated/models.py  — Pydantic v2 models
  - python/talonsight_ontology/generated/enums.py   — Literal types + const lists

Run: cd pcomjr-spec && python python/scripts/generate.py
"""

import json
import os
import textwrap
from pathlib import Path
from typing import Any

SCHEMA_PATH = Path(__file__).resolve().parent.parent.parent / "schema" / "pcomjr.schema.json"
OUT_DIR = Path(__file__).resolve().parent.parent / "talonsight_ontology" / "generated"


def load_schema() -> dict[str, Any]:
    with open(SCHEMA_PATH) as f:
        return json.load(f)


# ── Classification ───────────────────────────────────────────────────────────

def classify_defs(defs: dict[str, Any]) -> tuple[set[str], set[str], set[str], set[str]]:
    """Classify schema definitions into enums, unions, interfaces, artifact subtypes."""
    enum_types: set[str] = set()
    union_types: set[str] = set()
    interface_types: set[str] = set()
    artifact_subtypes: set[str] = set()

    for name, defn in defs.items():
        if defn.get("type") == "string" and "enum" in defn:
            enum_types.add(name)
        elif "oneOf" in defn and all("$ref" in o for o in defn["oneOf"]):
            union_types.add(name)
        elif defn.get("type") == "object" or "allOf" in defn:
            interface_types.add(name)
            if any(
                a.get("$ref") == "#/$defs/Artifact"
                for a in defn.get("allOf", [])
            ):
                artifact_subtypes.add(name)

    return enum_types, union_types, interface_types, artifact_subtypes


# ── Type Resolution ──────────────────────────────────────────────────────────

def resolve_ref(ref: str) -> str:
    return ref.replace("#/$defs/", "")


def to_snake(name: str) -> str:
    """Convert CamelCase to snake_case."""
    result = []
    for i, c in enumerate(name):
        if c.isupper() and i > 0:
            prev = name[i - 1]
            # Don't insert underscore between consecutive capitals unless next is lower
            if prev.islower() or (i + 1 < len(name) and name[i + 1].islower()):
                result.append("_")
        result.append(c.lower())
    return "".join(result)


def schema_to_python(prop: dict[str, Any], defs: dict[str, Any]) -> str:
    """Convert a JSON Schema property to a Python type annotation."""
    if "$ref" in prop:
        ref_name = resolve_ref(prop["$ref"])
        # If ref is an enum type, use the Literal alias
        if ref_name in _enum_types:
            return ref_name
        return f'"{ref_name}"'

    if "const" in prop:
        return f'Literal["{prop["const"]}"]'

    if "oneOf" in prop:
        parts = [schema_to_python(o, defs) for o in prop["oneOf"]]
        return " | ".join(parts)

    typ = prop.get("type")

    if typ == "string":
        if "enum" in prop:
            values = ", ".join(f'"{v}"' for v in prop["enum"])
            return f"Literal[{values}]"
        return "str"

    if typ in ("number", "integer"):
        if typ == "integer":
            return "int"
        return "float"

    if typ == "boolean":
        return "bool"

    if typ == "array":
        items = prop.get("items")
        if items:
            item_type = schema_to_python(items, defs)
            return f"list[{item_type}]"
        return "list[Any]"

    if typ == "object":
        if "properties" in prop and prop["properties"]:
            # Inline object — use dict for simplicity in generated code
            return "dict[str, Any]"
        additional = prop.get("additionalProperties")
        if additional is True or (isinstance(additional, dict) and additional):
            if isinstance(additional, dict) and "type" in additional:
                val_type = schema_to_python(additional, defs)
                return f"dict[str, {val_type}]"
            return "dict[str, Any]"
        return "dict[str, Any]"

    return "Any"


# ── Code Generation ──────────────────────────────────────────────────────────

# Module-level set populated by classify_defs for use in schema_to_python
_enum_types: set[str] = set()


def generate_enums(defs: dict[str, Any], enum_types: set[str]) -> str:
    """Generate enums.py — Literal type aliases and const lists."""
    lines: list[str] = [
        '"""',
        "PCOMJR Data Ontology — Generated Enum Types",
        "",
        "AUTO-GENERATED from schema/pcomjr.schema.json",
        "DO NOT EDIT MANUALLY — run `python scripts/generate.py` to regenerate.",
        '"""',
        "",
        "from typing import Literal",
        "",
    ]

    for name in sorted(enum_types):
        defn = defs[name]
        desc = defn.get("description", "")

        # Literal type alias
        values = ", ".join(f'"{v}"' for v in defn["enum"])
        lines.append(f"# {desc}" if desc else "")
        lines.append(f"{name} = Literal[{values}]")
        lines.append("")

        # Const list
        const_name = to_snake(name).upper()
        values_list = ", ".join(f'"{v}"' for v in defn["enum"])
        lines.append(f"{const_name}: list[{name}] = [{values_list}]")
        lines.append("")

    return "\n".join(lines)


def generate_models(
    defs: dict[str, Any],
    enum_types: set[str],
    union_types: set[str],
    interface_types: set[str],
    artifact_subtypes: set[str],
) -> str:
    """Generate models.py — Pydantic v2 BaseModel classes."""
    lines: list[str] = [
        '"""',
        "PCOMJR Data Ontology — Generated Pydantic Models",
        "",
        "AUTO-GENERATED from schema/pcomjr.schema.json",
        "DO NOT EDIT MANUALLY — run `python scripts/generate.py` to regenerate.",
        '"""',
        "",
        "from __future__ import annotations",
        "",
        "from datetime import datetime",
        "from typing import Any, Literal, Optional",
        "",
        "from pydantic import BaseModel, Field",
        "",
        "from .enums import (",
    ]

    # Import all enum types
    for name in sorted(enum_types):
        lines.append(f"    {name},")
    lines.append(")")
    lines.append("")

    # Union type aliases
    for name in sorted(union_types):
        defn = defs[name]
        desc = defn.get("description", "")
        members = " | ".join(resolve_ref(o["$ref"]) for o in defn["oneOf"])
        if desc:
            lines.append(f"# {desc}")
        lines.append(f"{name} = {members}")
        lines.append("")

    # Determine emit order: base types first, then subtypes
    # Collect simple interfaces (no allOf, not Decision)
    simple = sorted(
        name
        for name in interface_types
        if name not in artifact_subtypes and name != "Decision"
    )

    # Emit Artifact first (it's the base)
    artifact_order = ["Artifact"] + sorted(
        name for name in artifact_subtypes if name != "Decision"
    ) + ["Decision"]

    # Filter: only emit what exists
    simple = [n for n in simple if n != "Artifact"]

    # Emit simple interfaces
    for name in simple:
        _emit_model(lines, name, defs[name], defs, is_artifact_sub=False)

    # Emit Artifact and subtypes
    for name in artifact_order:
        if name not in defs:
            continue
        is_sub = name in artifact_subtypes or name == "Decision"
        _emit_model(lines, name, defs[name], defs, is_artifact_sub=is_sub and name != "Artifact")

    return "\n".join(lines)


def _emit_model(
    lines: list[str],
    name: str,
    defn: dict[str, Any],
    defs: dict[str, Any],
    is_artifact_sub: bool,
) -> None:
    """Emit a single Pydantic model class."""
    desc = defn.get("description", "")

    base_class = "Artifact" if is_artifact_sub else "BaseModel"
    lines.append("")
    lines.append(f"class {name}({base_class}):")
    if desc:
        lines.append(f'    """{desc}"""')
    lines.append("")

    # Collect all properties
    all_props: dict[str, Any] = {}
    all_required: set[str] = set(defn.get("required", []))

    if "properties" in defn:
        all_props.update(defn["properties"])

    # allOf merging
    for entry in defn.get("allOf", []):
        if "$ref" in entry:
            continue  # Skip base class ref
        if "properties" in entry:
            all_props.update(entry["properties"])
        for r in entry.get("required", []):
            all_required.add(r)

    # For subtypes, skip fields that are on the base Artifact
    skip_fields: set[str] = set()
    if is_artifact_sub:
        base_props = defs.get("Artifact", {}).get("properties", {})
        skip_fields = set(base_props.keys())

    emitted = False
    for prop_name, prop_def in all_props.items():
        # Skip inherited fields (unless they have a const override or narrow the type via $ref)
        if prop_name in skip_fields and "const" not in prop_def and "$ref" not in prop_def:
            continue

        py_type = schema_to_python(prop_def, defs)
        field_name = prop_name

        is_required = (prop_name in all_required and prop_name not in skip_fields) or (prop_name == "content" and "$ref" in prop_def)
        has_const = "const" in prop_def

        if has_const:
            # Const fields get a default
            const_val = prop_def["const"]
            field_desc = prop_def.get("description", "")
            if field_desc:
                lines.append(f'    {field_name}: {py_type} = Field(default="{const_val}", description="{field_desc}")')
            else:
                lines.append(f'    {field_name}: {py_type} = "{const_val}"')
        elif is_required:
            field_desc = prop_def.get("description", "")
            if field_desc:
                lines.append(f'    {field_name}: {py_type} = Field(..., description="{field_desc}")')
            else:
                lines.append(f"    {field_name}: {py_type}")
        else:
            field_desc = prop_def.get("description", "")
            default = prop_def.get("default")
            if default is not None:
                # Convert JSON literals to Python: false→False, true→True, null→None
                if isinstance(default, bool):
                    default_repr = "True" if default else "False"
                elif default is None:
                    default_repr = "None"
                else:
                    default_repr = json.dumps(default)
                if field_desc:
                    lines.append(f'    {field_name}: Optional[{py_type}] = Field(default={default_repr}, description="{field_desc}")')
                else:
                    lines.append(f"    {field_name}: Optional[{py_type}] = {default_repr}")
            else:
                if field_desc:
                    lines.append(f'    {field_name}: Optional[{py_type}] = Field(default=None, description="{field_desc}")')
                else:
                    lines.append(f"    {field_name}: Optional[{py_type}] = None")
        emitted = True

    if not emitted:
        lines.append("    pass")

    lines.append("")


def main():
    schema = load_schema()
    defs = schema["$defs"]

    global _enum_types
    enum_types, union_types, interface_types, artifact_subtypes = classify_defs(defs)
    _enum_types = enum_types

    OUT_DIR.mkdir(parents=True, exist_ok=True)

    # Generate enums
    enums_code = generate_enums(defs, enum_types)
    (OUT_DIR / "enums.py").write_text(enums_code)
    print(f"✓ Generated {OUT_DIR / 'enums.py'} ({len(enum_types)} enum types)")

    # Generate models
    models_code = generate_models(defs, enum_types, union_types, interface_types, artifact_subtypes)
    (OUT_DIR / "models.py").write_text(models_code)
    print(f"✓ Generated {OUT_DIR / 'models.py'} ({len(interface_types)} models)")

    # __init__.py for the generated subpackage
    init_code = '"""Auto-generated PCOMJR types. Do not edit."""\n\nfrom .enums import *  # noqa: F401,F403\nfrom .models import *  # noqa: F401,F403\n'
    (OUT_DIR / "__init__.py").write_text(init_code)
    print(f"✓ Generated {OUT_DIR / '__init__.py'}")


if __name__ == "__main__":
    main()
